src/fsharp/ConstraintSolver.fs

```diff
@@ -705,77 +708,80 @@ let CheckWarnIfRigid (csenv:ConstraintSolverEnv) ty1 (r:Typar) ty =
-           RepeatWhileD ndeep (fun ndeep -> SolveRelevantMemberConstraintsForTypar csenv ndeep false trace r)
-       else 
-           CompleteD) ++ (fun _ ->
+    trackErrors{
+        do! DepthCheck ndeep m
+        match ty1 with 
+        | TType_var r | TType_measure (Measure.Var r) ->
+            // The types may still be equivalent due to abbreviations, which we are trying not to eliminate 
+            if typeEquiv csenv.g ty1 ty then ()
+            else
+                // The famous 'occursCheck' check to catch "infinite types" like 'a = list<'a> - see also https://github.com/Microsoft/visualfsharp/issues/1170
+                if occursCheck csenv.g r ty then return! ErrorD (ConstraintSolverInfiniteTypes(csenv.eContextInfo, csenv.DisplayEnv, ty1, ty, m, m2))
+                else
+                    // Note: warn _and_ continue! 
+                    do! CheckWarnIfRigid csenv ty1 r ty
+                    // Record the solution before we solve the constraints, since 
+                    // We may need to make use of the equation when solving the constraints. 
+                    // Record a entry in the undo trace if one is provided 
+                    trace.Exec (fun () -> r.typar_solution <- Some ty) (fun () -> r.typar_solution <- None)
       
-      // Re-solve the other constraints associated with this type variable 
-      solveTypMeetsTyparConstraints csenv ndeep m2 trace ty r))
+                    (* dprintf "setting typar %d to type %s at %a\n" r.Stamp ((DebugPrint.showType ty)) outputRange m; *)
 
-    | _ -> failwith "SolveTyparEqualsType")
+                    // Only solve constraints if this is not an error var 
+                    if r.IsFromError then () 
+                    else
+                        // Check to see if this type variable is relevant to any trait constraints. 
+                        // If so, re-solve the relevant constraints. 
+                        if csenv.SolverState.ExtraCxs.ContainsKey r.Stamp then 
+                            do! RepeatWhileD ndeep (fun ndeep -> SolveRelevantMemberConstraintsForTypar csenv ndeep false trace r)
+                        // Re-solve the other constraints associated with this type variable 
+                        return! solveTypMeetsTyparConstraints csenv ndeep m2 trace ty r
+
+        | _ -> failwith "SolveTyparEqualsType"
+    }
     
 
 /// Apply the constraints on 'typar' to the type 'ty'
 and solveTypMeetsTyparConstraints (csenv:ConstraintSolverEnv) ndeep m2 trace ty (r: Typar) = 
     let g = csenv.g
-    // Propagate compat flex requirements from 'tp' to 'ty'
-    SolveTypIsCompatFlex csenv trace r.IsCompatFlex ty ++ (fun () -> 
-    // Propagate dynamic requirements from 'tp' to 'ty'
-    SolveTypDynamicReq csenv trace r.DynamicReq ty ++ (fun () -> 
-    // Propagate static requirements from 'tp' to 'ty' 
-    SolveTypStaticReq csenv trace r.StaticReq ty ++ (fun () -> 
+    trackErrors {
```

Comment:
Could you change the style used throughout to 

```fsharp
and solveTypMeetsTyparConstraints (csenv:ConstraintSolverEnv) ndeep m2 trace ty (r: Typar) = trackChanges {
    let g = csenv.g
    ...
  }

and ...
```
I believe this will greatly reduce the size of this diff and the code churn